[Security Solution] New Side nav integrating links config

semd opened the pull request:
## Summary

Integrating new security links configuration into the new navigation. This is enabling us to implement:

- Selected item blue highlight
- "Manage" panel sub-categories
- New deepLinks hierarchy
- Navigation footer items
- appLinks updater subscriptions
 
![side_nav_screenshot](https://user-images.githubusercontent.com/17747913/168800725-70160a85-9977-484d-9733-145cd3710fc5.png)

Demo:

https://user-images.githubusercontent.com/17747913/168801198-16dcb2fa-e72d-4b3d-baa8-2ade38aa7114.mov

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

machadoum committed:
Update navigation landing pages to use appLinks config

semd committed:
align app links changes

semd committed:
merge main

semd committed:
link configs refactor to use updater$

semd committed:
merge

semd committed:
navigation panel categories

semd committed:
test and type fixes

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

semd committed:
types changes

semd committed:
Merge branch '130023_security_side_nav_config' of https://github.com/semd/kibana into 130023_security_side_nav_config

semd committed:
shared style change moved to a separate PR

semd committed:
use old deep links

semd committed:
Merge remote-tracking branch 'upstream/main' into 130023_security_side_nav_config

semd committed:
minor changes after ux meeting

semd committed:
conflicts solved

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

semd committed:
add links filtering

semd committed:
remove duplicated categories

semd committed:
temporary increase of plugin size limit

semd committed:
swap management links order

semd committed:
improve performance closing nav panel

semd committed:
test updated

semd commented on packages/kbn-optimizer/limits.yml:
```diff
@@ -93,7 +93,7 @@ pageLoadAssetSize:
   interactiveSetup: 80000
   expressionTagcloud: 27505
-  securitySolution: 273763
+  securitySolution: 289364 # This is temporary: https://github.com/elastic/kibana/pull/132210 update-limits when feature flag removed and deprecated code cleaned
```
@elastic/kibana-operations We are changing the side navigation components, but It is under a feature & uiSettings flag, so we have to maintain both navigations for a while, having duplicated components causes the plugin size to rise significantly.
The size will come back down when the conditional is removed and we are able to clean the old navigation components.
Do we need to do anything else in this situation?

mistic replied:
I think we are okay for now with that until we can remove it and get the bundles downsized again.

dasansol92 commented on x-pack/plugins/security_solution/public/management/links.ts:
```diff
@@ -175,8 +174,8 @@ export const navigationCategories: NavigationCategories = [
       SecurityPageName.eventFilters,
-      SecurityPageName.blocklist,
       SecurityPageName.hostIsolationExceptions,
+      SecurityPageName.blocklist,
```
Thanks for sorting this!

Also, seems the hostIsolationException item is not disabled when env does not match the criteria. We already have this logic in the existing navigation [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/common/components/navigation/use_security_solution_navigation/use_navigation_items.tsx#L119-L121)

semd replied:
Excluded the link using the hook here: 
https://github.com/elastic/kibana/pull/132210/files#diff-d1acbf8273add55b64dd3840a1e7eca4a36c4baa3f4e5b31b86d73bd93cecd0fR143-R148

We should change this approach at some point, filtering all links when the plugin mounts.

machadoum commented on x-pack/plugins/security_solution/public/common/components/navigation/security_side_nav/security_side_nav.tsx:
```diff
@@ -0,0 +1,152 @@
+  href: string;
+  onClick: MouseEventHandler;
+}> = ({ isSelected, title, href, onClick }) => (
+  // eslint-disable-next-line @elastic/eui/href-or-on-click
```
nit: I like to use the `LinkAnchor`component to avoid disabling the linter.

mistic reviewed (approved):
Changes under operations team code owners LGTM

angorayc commented on x-pack/plugins/security_solution/public/common/components/navigation/solution_grouped_nav/solution_grouped_nav.tsx:
```diff
@@ -163,5 +149,84 @@ export const SolutionGroupedNavComponent: React.FC<SolutionGroupedNavProps> = ({
+
+  return (
+    // eslint-disable-next-line @elastic/eui/href-or-on-click
+    <EuiLink
```
LinkAnchor component ++

semd replied:
good point, the thing is that I tried to use only generic components in the `solution_grouped_nav`. Just in case we want to make it available to other solutions later on, in the shared-ux plugin.

angorayc commented on x-pack/plugins/security_solution/public/common/components/navigation/solution_grouped_nav/solution_grouped_nav_panel.tsx:
```diff
@@ -105,5 +116,61 @@ const SolutionGroupedNavPanelComponent: React.FC<SolutionGroupedNavPanelProps> =
+        }, []);
+
+        return (
+          <Fragment key={label}>
```
nit: The key here is coming from i18n.translate, might be better to actually assign them a key prop in x-pack/plugins/security_solution/public/management/links.ts and use that prop?

angorayc commented on x-pack/plugins/security_solution/public/common/components/navigation/solution_grouped_nav/solution_grouped_nav_panel.tsx:
```diff
@@ -105,5 +116,61 @@ const SolutionGroupedNavPanelComponent: React.FC<SolutionGroupedNavPanelProps> =
+    {items.map(({ id, href, onClick, label, description }) => (
+      <Fragment key={id}>
+        <EuiDescriptionListTitle>
+          <a
```
Is there any particular reason we use <a> here?

semd replied:
Just because it has to take the styles from `EuiDescriptionListTitle`, using `EuiLink` inside would add some more styles that are not needed.

angorayc reviewed (approved):
Great work Sergi, tested locally with a normal user and a no cases privilege user

machadoum commented on x-pack/plugins/security_solution/public/common/links/types.ts:
```diff
@@ -53,24 +72,33 @@ export interface LinkItem {
+  sideNavDisabled?: boolean;
+  /**
+   * Enables link in the side navigation. Defaults to false.
+   */
```
First, thank you very much for adding comments. It is very helpful. 

I think this comment is wrong. `skipUrlState` disabled the query string in the URL.

semd committed:
host isolation page filterd and some improvements

YulNaumenko commented on x-pack/plugins/security_solution/public/plugin.tsx:
```diff
@@ -190,7 +191,7 @@ export class Plugin implements IPlugin<PluginSetup, PluginStart, SetupPlugins, S
   }
 
-  public start(core: CoreStart, plugins: StartPlugins) {
+  public async start(core: CoreStart, plugins: StartPlugins) {
```
We need to avoid asynchronous methods for the plugin start and setup according to the guidelines.

semd committed:
remove async from plugin start

semd committed:
move links register from start to mount

semd committed:
restore size limits

semd committed:
merge main

YulNaumenko reviewed (approved):
LGTM. I have some minor UX issues, but it could be fixed later:
<img width="657" alt="Screen Shot 2022-05-17 at 12 58 07 PM" src="https://user-images.githubusercontent.com/55110838/169045441-bb169682-fe2a-4ce7-8db6-26c6b8e28b0c.png">

This highlighting box appears only sometimes:

https://user-images.githubusercontent.com/55110838/169045615-ea63031c-fc5a-4231-a1f5-314176b0f8f3.mov

About increasing the initial plugin bundle size, I'm wonder why this is happened: do we use a lazy loading approach for the components?

machadoum committed:
Merge remote-tracking branch 'elastic/main' into 130023_security_side_nav_config

machadoum commented:
Hey @YulNaumenko, Sergi fixed the bundle size issue by lazy loading  `common/links/app_links`. He also has restored the original limit on this commit https://github.com/elastic/kibana/pull/132210/commits/c49f7311165ca8213eb1712709db68c3be4551f3.

For documentation:
Moving the app registering to mount (ac944aa01a8dda41c46d84ccc584b3a6181d6cdf) has introduced a bug. The security solution link doesn't appear on the kibana home page and global menu. 
<img width="1775" alt="Screenshot 2022-05-19 at 10 13 45" src="https://user-images.githubusercontent.com/1490444/169255243-22bd8193-caff-4587-8f9a-b7b812b12eb0.png">

machadoum committed:
Fix use_show_timeline unit tests

YulNaumenko committed:
reverted changes where we moved the navigation registration to the component mount - this caused the problem when we do not start from the security app we never see that in the navigation

YulNaumenko committed:
Merge branch '130023_security_side_nav_config' of https://github.com/semd/kibana into 130023_security_side_nav_config

YulNaumenko committed:
Merge branch 'main' into 130023_security_side_nav_config

parkiino reviewed (approved):
checked it out and saw that the management policy settings stuff looks fine!

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

machadoum committed:
Fix HostIsolationExceptions link visibility

The Problem:
"In the scenario user does not have isolate privileges but there is at least one host isolation exception, this hook will return false at the beginning and true after the data has been loaded."
https://github.com/elastic/kibana/pull/132210#discussion_r877177642

Temporary fix:
Delete code that removes HostIsolationExceptions link from Applinks and only remove the link from Security navigation.

We will investigate a better solution to remove HostIsolationExceptions from AppLinks.

machadoum commented on x-pack/plugins/security_solution/public/common/components/navigation/security_side_nav/security_side_nav.tsx:
```diff
@@ -0,0 +1,155 @@
+  const selectedId = useSelectedId();
+
+  // TODO: move this exclusion logic to getManagementLinkItems
+  const canSeeHostIsolationExceptions = useCanSeeHostIsolationExceptionsMenu();
```
Hey @dasansol92, As suggested by @semd we are rolling back the change that removes `HostIsolationExceptions` from AppLinks. Here is the commit 28b90678ff5efea7f9beda2bbab36a4fc450b316

dasansol92 replied:
Awesome, so it will be hidden by default and displayed if needed right?

machadoum replied:
It will render on the security solution menu and manage the landing page when `useCanSeeHostIsolationExceptionsMenu` return true and hidden when it returns false. It should work similarly to the old menu: https://github.com/elastic/kibana/blob/a0cf4506d1806af7a9380ff73f681bc611126d4b/x-pack/plugins/security_solution/public/common/components/navigation/use_security_solution_navigation/use_navigation_items.tsx#L119

machadoum committed:
Fix manageOldSiemRoutes been called before deepLinks initialization
